### ISSUE TITLE
[Tests] munmap LTP test succeeds unexpectedly

### ISSUE THREAD

**Issue by prp** [lsds/sgx-lkl#611]
We observe the following failure for `kernel-syscalls-munmap-munmap03-(nonrelease)-(run-hw)-(8-ethreads)`:
```
[   SGX-LKL  ] SGX-LKL (OE) 0.2.0-dev (cc81a85) LKL 5.3.0 [NON-RELEASE build (-O3)] [HARDWARE DEBUG]
[   SGX-LKL  ] Warning: SGXLKL_TRACE_SIGNAL ignored in non-debug mode.
[   SGX-LKL  ] nproc=4 ETHREADS=8 CMDLINE="mem=512m"
[   SGX-LKL  ] HW TLS support: conf->fsgsbase=1
[   SGX-LKL  ] Registering disk 0 (path='sgxlkl-miniroot-fs.img', mnt='/', [RW    ])
[   SGX-LKL  ] No tap device specified, networking will not be available.
[   SGX-LKL  ] get_signed_libsgxlkl_path... result=/opt/sgx-lkl-nonrelease/bin/../lib/libsgxlkl.so.signed
[   SGX-LKL  ] oe_create_enclave... result=0 (OE_OK)
[   SGX-LKL  ] sgxlkl_enclave_init(ethread_id=0)
[    0.000000] Linux version 5.3.0+ (AzDevOps@cisca24ba000207) (gcc version 7.5.0 (Ubuntu 7.5.0-3ubuntu1~18.04)) #1 Wed Jul 8 10:28:37 UTC 2020
[    0.000000] memblock address range: 0x7f2b207ec000 - 0x7f2b407ec000
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 129280
[    0.000000] Kernel command line: mem=512m console=hvc0 virtio_mmio.device=256@0x1000000:1
[    0.000000] Dentry cache hash table entries: 65536 (order: 7, 524288 bytes, linear)
[    0.000000] Inode-cache hash table entries: 32768 (order: 6, 262144 bytes, linear)
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory available: 516256k/524288k RAM
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 4096
[    0.000000] lkl: irqs initialized
[    0.000000] clocksource: lkl: mask: 0xffffffffffffffff max_cycles: 0x1cd42e4dffb, max_idle_ns: 881590591483 ns
[    0.000000] lkl: time and timers initialized (irq2)
[    0.000514] printk: console [hvc0] enabled
[    0.000514] pid_max: default: 4096 minimum: 301
[    0.000514] Mount-cache hash table entries: 1024 (order: 1, 8192 bytes, linear)
[    0.000514] Mountpoint-cache hash table entries: 1024 (order: 1, 8192 bytes, linear)
[    0.002169] devtmpfs: initialized
[    0.004811] random: get_random_bytes called from 0x00007f2b00594e59 with crng_init=0
[    0.004811] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.004811] futex hash table entries: 16 (order: -4, 384 bytes, linear)
[    0.004811] xor: automatically using best checksumming function   8regs     
[    0.004811] NET: Registered protocol family 16
[    0.007030] clocksource: Switched to clocksource lkl
[    0.007030] NET: Registered protocol family 2
[    0.007590] tcp_listen_portaddr_hash hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.007590] TCP established hash table entries: 4096 (order: 3, 32768 bytes, linear)
[    0.007590] TCP bind hash table entries: 4096 (order: 3, 32768 bytes, linear)
[    0.007590] TCP: Hash tables configured (established 4096 bind 4096)
[    0.007590] UDP hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.007590] UDP-Lite hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.008150] NET: Registered protocol family 1
[    0.008150] virtio-mmio: Registering device virtio-mmio.0 at 0x1000000-0x10000ff, IRQ 1.
[    0.008150] workingset: timestamp_bits=62 max_order=18 bucket_order=0
[    0.009262] NET: Registered protocol family 38
[    0.009262] io scheduler mq-deadline registered
[    0.009262] io scheduler kyber registered
[    0.105885] software IO TLB: mapped [mem 0x7f2bb3be8000-0x7f2bb7be8000] (64MB)
[    0.109045] loop: module loaded
[    0.117483] device-mapper: ioctl: 4.40.0-ioctl (2019-01-18) initialised: dm-devel@redhat.com
[    0.117988] wireguard: WireGuard 0.0.20191219 loaded. See www.wireguard.com for information.
[    0.117988] wireguard: Copyright (C) 2015-2019 Jason A. Donenfeld <Jason@zx2c4.com>. All Rights Reserved.
[    0.118495] NET: Registered protocol family 10
[    0.119002] Segment Routing with IPv6
[    0.119002] sit: IPv6, IPv4 and MPLS over IPv4 tunneling driver
[    0.120670] This architecture does not have kernel memory protection.
[    0.125121] Run /init as init process
[    0.125692] random: crng init done
[[  SGX-LKL ]] wg0 has public key cYHJe0oAwxyQWgdDb1rRvT0Gk0ynJPQNBBAtzcqu9E8=
[    0.127954] virtio_blk virtio1: [vda] 3072000 512-byte logical blocks (1.57 GB/1.46 GiB)
[    0.135232] EXT4-fs (vda): mounted filesystem with ordered data mode. Opts: 
munmap03    1  [1;31mTFAIL[0m  :  munmap03.c:105: munmap succeeded unexpectedly
munmap03    2  [1;31mTFAIL[0m  :  munmap03.c:105: munmap succeeded unexpectedly
munmap03    3  [1;31mTFAIL[0m  :  munmap03.c:105: munmap succeeded unexpectedly
[    0.217208] reboot: Restarting system
[   SGX-LKL  ] ethread (6: 0) ethread (7: 0) ethread (2: 0) ethread (1: 0) init (0: 0 exit=1) ethread (5: 0) ethread (4: 0) ethread (3: 0) 
[   SGX-LKL  ] oe_terminate_enclave... done
[   SGX-LKL  ] SGX-LKL-OE exit: exit_status=1
TEST_FAILED EXIT CODE: 1
```
The reason for this should be investigated.

(cc: @davidchisnall)

**Comment by hukoyu** [comment]
@shaikshavali1 can you investigate this please.

**Comment by shaikshavali1** [comment]
@prp, I tried to reproduce this issue, but not successful in reproducing it. 
Is this issue observed consistently/randomly?

(cc: @hukoyu )

**Comment by prp** [comment]
@shaikshavali1 So far I've only seen this once. Let's keep this issue open while we see if the problems appears again in other CI runs.

**Comment by davidchisnall** [comment]
This test should be failing 100% of the time.  Our munmap implementation unconditionally returns success.  I don't see how this can ever pass.

**Comment by hukoyu** [comment]
@davidchisnall this is a passing run log:
```
[[  SGX-LKL ]] lkl_mount_disks(): Set working directory: /
[[  SGX-LKL ]] libc_start_main_stage2(): Calling app main: /ltp/testcases/kernel/syscalls/munmap/munmap03
munmap03    1  [1;32mTPASS[0m  :  failed as expected: TEST_ERRNO=EINVAL(22): Invalid argument
munmap03    2  [1;32mTPASS[0m  :  failed as expected: TEST_ERRNO=EINVAL(22): Invalid argument
munmap03    3  [1;32mTPASS[0m  :  failed as expected: TEST_ERRNO=EINVAL(22): Invalid argument
[[  SGX-LKL ]] lkl_terminate(): terminating LKL (exit_status=0)
[[  SGX-LKL ]] lkl_termination_thread(): termination thread unblocked
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_sys_chdir(/)
[[  SGX-LKL ]] display_mount_table(): ========= /proc/mounts ===========
/dev/vda / ext4 rw,relatime 0 0
devtmpfs /dev devtmpfs rw,relatime 0 0
tmpfs /dev/shm tmpfs rw,relatime 0 0
tmpfs /tmp tmpfs rw,relatime,mode=777 0 0
tmpfs /mnt tmpfs rw,relatime,mode=777 0 0
none /sys sysfs rw,relatime 0 0
tmpfs /run tmpfs rw,relatime,mode=700 0 0
proc /proc proc rw,relatime 0 0
[[  SGX-LKL ]] display_mount_table(): ==================================
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_umount_timeout("/", MNT_DETACH, 2000)
[[  SGX-LKL ]] display_mount_table(): /proc/mounts cannot be accessed
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_virtio_netdev_remove()
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_sys_halt()
[    0.189884] reboot: Restarting system
[[  SGX-LKL ]] lkl_termination_thread(): lthread_detach2() done
[   SGX-LKL  ] init (0: 0 exit=0) ethread (7: 0) ethread (4: 0) ethread (1: 0) ethread (5: 0) ethread (2: 0) ethread (6: 0) ethread (3: 0) 
[   SGX-LKL  ] oe_terminate_enclave... done
[   SGX-LKL  ] SGX-LKL-OE exit: exit_status=0
```

**Comment by davidchisnall** [comment]
If you can reproduce a passing log, please step through the munmap call and find where the success return is coming from.  [The function that implements that syscall for us](https://github.com/lsds/sgx-lkl/blob/7802da2876944967938d2155a83bca04d4faf2a9/src/enclave/enclave_mem.c#L167) returns success unconditionally.  If we are seeing a failure result, we are hitting some very surprising code paths.

**Comment by hukoyu** [comment]
@shaikshavali1 can you provide info David is asking for.

**Comment by shaikshavali1** [comment]
@davidchisnall, This test case performs below sub test cases and expects EINVAL from munmap.
1. Pass some random address. 
2. Pass length as 0.
3. Pass lesser length than mapped size. 

This test case checks both [return value and error number after returning from the system](https://github.com/lsds/sgx-lkl/blob/9dd6fb9213d9defb3c492f4a29ded2c3158fb838/tests/ltp/patches/munmap03.patch#L14). This error number is updated to EINVAL in "[enclave_mmap](https://github.com/lsds/sgx-lkl/blob/7802da2876944967938d2155a83bca04d4faf2a9/src/enclave/enclave_mem.c#L446)".  I tried to reproduce the issue (which doesn't set the errno to EINVAL) but it is not reproducible.

```
Thread 6 "ENCLAVE" hit Breakpoint 1, enclave_munmap (addr=0xffffffffffffffff, length=8192) at enclave/enclave_mem.c:440
440         size_t pages = DIV_ROUNDUP(length, PAGE_SIZE);
(gdb) bt
#0  enclave_munmap (addr=0xffffffffffffffff, length=8192) at enclave/enclave_mem.c:440
#1  0x00007fff000e1de7 in syscall_SYS_munmap (addr=<optimized out>, length=<optimized out>) at enclave/enclave_mem.c:184
#2  0x00007fff000a55bc in munmap ()
#3  0x0000000000002000 in ?? ()
#4  0xffffffffffffffff in ?? ()
#5  0x00007fff3de4da2e in ?? ()
#6  0xffffffffffffffff in ?? ()
#7  0xffffffffffffffff in ?? ()
#8  0x00007fff40a6cec0 in ?? ()
#9  0x0000000000000000 in ?? ()
(gdb) c
Continuing.
munmap03    1  TPASS:  failed as expected: TEST_ERRNO=EINVAL(22): Invalid argument

Thread 6 "ENCLAVE" hit Breakpoint 1, enclave_munmap (addr=0x7fff3de47000, length=0) at enclave/enclave_mem.c:440
440         size_t pages = DIV_ROUNDUP(length, PAGE_SIZE);
(gdb) s
439     {
(gdb) s
440         size_t pages = DIV_ROUNDUP(length, PAGE_SIZE);
(gdb) s
443         if ((uintptr_t)addr % PAGE_SIZE != 0 || length == 0 ||
(gdb) s
439     {
(gdb) s
443         if ((uintptr_t)addr % PAGE_SIZE != 0 || length == 0 ||
(gdb) s
446             errno = EINVAL;
(gdb) p addr
$1 = (void *) 0x7fff3de47000
(gdb) p length
$2 = <optimized out>
(gdb) s
447             return -1;
(gdb) s
483     }
(gdb) s
syscall_SYS_munmap (addr=<optimized out>, length=<optimized out>) at enclave/enclave_mem.c:186
186     }
(gdb) s
0x00007fff000a55bc in munmap ()
(gdb) c
Continuing.
munmap03    2  TPASS  :  failed as expected: TEST_ERRNO=EINVAL(22): Invalid argument

Thread 6 "ENCLAVE" hit Breakpoint 1, enclave_munmap (addr=0x7fff3de47001, length=4096) at enclave/enclave_mem.c:440
440         size_t pages = DIV_ROUNDUP(length, PAGE_SIZE);
(gdb) s
439     {
(gdb) s
440         size_t pages = DIV_ROUNDUP(length, PAGE_SIZE);
(gdb) s
443         if ((uintptr_t)addr % PAGE_SIZE != 0 || length == 0 ||
(gdb) list
438     int enclave_munmap(void* addr, size_t length)
439     {
440         size_t pages = DIV_ROUNDUP(length, PAGE_SIZE);
441
442         // Make sure addr is page aligned and size is greater than 0
443         if ((uintptr_t)addr % PAGE_SIZE != 0 || length == 0 ||
444             !in_mmap_range(addr, length))
445         {
446             errno = EINVAL;
447             return -1;
(gdb) s
439     {
(gdb) s
443         if ((uintptr_t)addr % PAGE_SIZE != 0 || length == 0 ||
(gdb) s
446             errno = EINVAL;
(gdb) list
441
442         // Make sure addr is page aligned and size is greater than 0
443         if ((uintptr_t)addr % PAGE_SIZE != 0 || length == 0 ||
444             !in_mmap_range(addr, length))
445         {
446             errno = EINVAL;
447             return -1;
448         }
449
450         size_t index = addr_to_index(addr);
(gdb) s
447             return -1;
(gdb) s
483     }
(gdb) s
syscall_SYS_munmap (addr=<optimized out>, length=<optimized out>) at enclave/enclave_mem.c:186
186     }
(gdb) list
181             lt->attr.stack_size = length;
182             return 0;
183         }
184         enclave_munmap(addr, length);
185         return 0;
186     }
187
188     int syscall_SYS_msync(void* addr, size_t length, int flags)
189     {
190         return 0;
(gdb) s
0x00007fff000a55bc in munmap ()
(gdb) bt
#0  0x00007fff000a55bc in munmap ()
#1  0x00007fff3de47001 in ?? ()
#2  0x0000000000001000 in ?? ()
#3  0x00007fff3de4d9be in ?? ()
#4  0x0000000000000000 in ?? ()
(gdb) s
Single stepping until exit from function munmap,
which has no line number information.
Warning:
Cannot insert breakpoint 0.
Cannot access memory at address 0x0

0x00007fff3de4d9be in ?? ()
(gdb) c
Continuing.
munmap03    3  TPASS  :  failed as expected: TEST_ERRNO=EINVAL(22): Invalid argument

Thread 6 "ENCLAVE" hit Breakpoint 1, enclave_munmap (addr=0x7fff3de47000, length=8192) at enclave/enclave_mem.c:440
440         size_t pages = DIV_ROUNDUP(length, PAGE_SIZE);
(gdb)

```

**Comment by davidchisnall** [comment]
`enclave_mmap` should not be setting `errno`, that's a layering violation.  #597 removes that behaviour.  The syscall wrapper will set errno based on the return value of the syscall, so setting `errno` in the enclave layer is a problem anyway.

**Comment by SeanTAllen** [comment]
This now fails consistently on #604 which is an improvement. I'm now working on this issue as part of #604.

**Comment by SeanTAllen** [comment]
So at minimum, the patch that is being applied to this test is wrong.

It was made to attempt to get the test to pass but, unfortunately, the code in question is wrong.
It is always giving a return value of 0 regardless of if munmap succeeded. That's a bug. But the patch tried to adjust for it.

I've fixed that bit and am working on getting it completely correct.

**Comment by SeanTAllen** [comment]
#641 removes the incorrect test patch. Without that patch, this will always fail. Closing this in favor of #642.
